src/misc/paper-write.ts

```diff
@@ -0,0 +1,244 @@
+import {
+  EquipmentSlot,
+  ItemStack,
+  PlayerInventory,
+} from 'org.bukkit.inventory';
+import { PlayerInteractEvent } from 'org.bukkit.event.player';
+import { Player } from 'org.bukkit.entity';
+import { color, text } from 'craftjs-plugin/chat';
+import { ChatColor } from 'net.md_5.bungee.api';
+import { isRightClick } from '../common/helpers/click';
+import { ChatMessage, GLOBAL_PIPELINE } from '../chat/pipeline';
+import { CustomItem } from '../common/items/CustomItem';
+import { VkItem } from '../common/items/VkItem';
+import * as yup from 'yup';
+import { isBuffer } from 'lodash';
+
+// Max length of text.
+const MAX_LENGTH = 128;
+
+const playersWriting: Player[] = [];
+
+/**
+ * CustomModelData table of Paper
+ * id 0: Normal paper
+ * 1: Written paper
+ * 2: Sealed paper
+ * 3: Empty envelope
+ * 4: Envelope with letter
+ * 5: Closed/sealed envelope
+ */
+
+const Envelope = new CustomItem({
+  id: 3,
+  name: 'Kirjekuori',
+  type: VkItem.PAPER,
+  modelId: 3,
+});
+
+export const EnvelopeWithLetter = new CustomItem({
+  id: 4,
+  name: 'Kirjekuori',
+  type: VkItem.PAPER,
+  modelId: 4,
+  data: {
+    letter: yup.string().required(),
+  },
+});
+
+export const EnvelopeSealed = new CustomItem({
+  id: 5,
+  name: 'Suljettu Kirjekuori',
+  type: VkItem.PAPER,
+  modelId: 5,
+  data: {
+    letter: yup.string().required(),
+  },
+});
+
+/**
+ * Writing on paper
+ */
+
+registerEvent(PlayerInteractEvent, async (event) => {
+  if (!isRightClick(event.action)) return;
+  if (event.hand !== EquipmentSlot.HAND) return;
+  if (event.item?.type !== Material.PAPER) return;
+  // Check if player has feather
+  const inventory = event.player.inventory as PlayerInventory;
+  const offHand = inventory.itemInOffHand;
+  if (offHand.type !== Material.FEATHER) return;
+  if (playersWriting.includes(event.player)) return;
+  // Check if the paper already has something written on it
+  if (event.item.itemMeta.lore) {
+    event.player.sendActionBar('§7Tälle paperille on jo kirjoitettu jotakin.');
+    return;
+  }
+  if (!event.player.inventory.contains(Material.INK_SAC, event.item.amount)) {
+    event.player.sendActionBar('§7Sinulla ei ole mustetta tarpeeksi.');
+    return;
+  }
+  event.player.sendMessage(
+    color(ChatColor.GRAY, text('Kirjoita haluamasi teksti chattiin!')),
+  );
+  playersWriting.push(event.player);
+});
+
+/**
+ * Reading paper contents
+ */
+registerEvent(PlayerInteractEvent, async (event) => {
+  if (!isRightClick(event.action)) return;
+  if (event.hand !== EquipmentSlot.HAND) return;
+  if (event.item?.type !== Material.PAPER) return;
+  // Check if player has feather
+  const inventory = event.player.inventory as PlayerInventory;
+  const offHand = inventory.itemInOffHand;
+  if (offHand.type == Material.FEATHER) return;
+  if (playersWriting.includes(event.player)) return;
+  // Check if the paper already has something written on it
+  if (event.item.itemMeta.customModelData != 1) return;
```

Comment:
CustomItem for modelid 1 could be used here? Prefer customitems whenever possible